lib/cpp/export/ecl/ostream.hpp

```diff
@@ -2,190 +2,170 @@
+ {
+    
+    template< typename stream >
+    stream& endl(stream &ios)
+    {
+        ios.put('\n');
+        ios.put('\r');
+    
+        // TODO: flush stream when appropriate functionality
+        // will be ready
+        return ios;
     }
-
-    int delim = higher_multiplicand;
-
-    if (value < 0) {
-        m_device->write((uint8_t *) "-", 1);
-        value *= -1;
-        delim *= -1;
-    }
-
-    for (unsigned int i = 1; i <= higher_multiplicand; i *= 10) {
-        out_digit = value / delim;
-        out_character = out_digit + 48;
-
-        if (m_device->write(reinterpret_cast< uint8_t* >(&out_character), 1) < 0) {
-            break; //FIXME: add error handling
-        }
-
-        value = value - out_digit * delim;
-        delim /= 10;
-    }
-
+ 
+ template< class IO_device >
+ class ostream
+ {
+ public:
+     // Provides type information of the underlying device
+     using device_type = IO_device;
+ 
+     // Initializes a stream with given device
+     // NOTE: device must be initialized and opened already
+     ostream(IO_device *device);
+     ~ostream();
+ 
+     ostream &operator<<(int value);
+     ostream &operator<<(unsigned int value);
+     ostream &operator<<(char character);
+     ostream &operator<<(const char *string);
+ 
+     // For I\O manipulators
+     ostream &operator<<(ostream& (*func)(ostream< IO_device >&));
+     // NOTE: this will be used later with different manipulators,
+     // such that used for hex or octal output of integers
+ 
+     // Puts a single character
+     ostream &put(char c);
+ 
+     // Disabled for now.
+     ostream &operator=(ostream &) = delete;
+     ostream(const ostream &) = delete;
+ 
+ private:
+     // Simply, a device driver object
+     IO_device *m_device;
+ };
+ 
+ 
+ //------------------------------------------------------------------------------
+ 
+ 
+ template< class IO_device >
+ ostream< IO_device >::ostream(IO_device *device)
+     :m_device{device}
+ {
+ }
+ 
+ template< class IO_device >
+ ostream< IO_device >::~ostream()
+ {
+ }
+ 
+ template< class IO_device >
+ ostream<IO_device> &ostream< IO_device >::operator<< (int value)
+ {
+     char buff[32];
+     
+     int n =  snprintf(buff, sizeof(buff), "%d", value);
+     
+     ecl_assert(n > 0 && n < static_cast<int>(sizeof(buff)));
+     
+     m_device->write(reinterpret_cast<uint8_t*>(buff), n);
+     
+     return *this;
+ }
+ 
+ 
+ template< class IO_device >
+ ostream<IO_device> &ostream< IO_device >::operator<< (unsigned int value)
+ {
+     
+    char buff[32];
+    
+    int n = snprintf(buff, sizeof(buff), "%u", value);
+    
+    ecl_assert(n > 0 && n < static_cast<int>(sizeof(buff)));
+
+    m_device->write(reinterpret_cast<uint8_t*>(buff), n);
+        
     return *this;
-}
+ }
+ 
+ 
+ template< class IO_device >
+ ostream<IO_device> &ostream< IO_device >::operator<<(char character)
+ {
+     
+    char buff [32];
+    
+    char n = snprintf(buff, sizeof(buff), "%c", character);
```

Comment:
No need to write character in buffer and then put it in the device. Let's just call `put()` here.